### ISSUE TITLE
'guicursor': Altering cursor in tmux + konsole alters it in all tmux panes/windows

### ISSUE THREAD

**Issue by ZyX-I** [neovim/neovim#6778]
- `nvim --version`: bde46fdec
- Terminal name/version: konsole-16.12.3, tmux-2.2
- `$TERM`: screen-256color

### Steps to reproduce

1. Run `tmux new-session \; split-window`
2. Run `nvim -u NONE -i NONE`.
3. Enter insert mode. Cursor will change to vertical line.
4. Given `<TmuxPrefix>` is tmux prefix (`<C-b>` by default) run `<TmuxPrefix>:select-pane -t :.+`.

### Actual behaviour

After 4 you are in another pane where Neovim was not opened, but cursor is a vertical line. Note that if it was opened, but was not in insert mode then cursor will still be wrong.

### Expected behaviour

It is still a block like it was before Neovim was run, vertical line only in tmux pane.

---

I actually do not like Neovim messing up with the cursor by default at all, empty `guicursor` by default would be better (init.vim is not an option, not for a developer since I need to use `-u NONE` often; previous variant with environment variable was better). But this behaviour is simply wrong, *do not try bypassing tmux*, disable changing cursor shape when in tmux instead, switching cursor shape should be a tmux responisibility. Current variant works well only if tmux user uses one pane and one window.

**Comment by justinmk** [comment]
> empty guicursor by default would be better

Adhering to the status quo because of some terminals' limitations prevents us from making the experience better for 80% of users, forever. Let's stop serving the lowest common denominator of terminals.

If we need to do some workaround for tmux and/or konsole, fine. I seem to recall @blueyed reporting something like this years ago, I thought there was some workaround, maybe related to this advice in `:h 'guicursor'`:

    set -ga terminal-overrides ',*:Ss=\E[%p1%d q:Se=\E[2 q'

**Comment by ZyX-I** [comment]
> Adhering to the status quo because of some terminals' limitations prevents us from making the experience better for 80% of users, forever. Let's stop serving the lowest common denominator of terminals.

@justinmk This part has nothing to do with terminal limitations. I do not like Neovim altering my cursor and it does so by default when I need `-u NONE` often and thus can’t set something in init.vim. Changing shape the way they are changed now is a matter of personal preference, I highly doubt current default is the best option for as much as 80% of users.

**Comment by justinmk** [comment]
The terminal limitation is that cursor isn't queryable. 

>  it does so by default when I need -u NONE often and thus can’t set something in init.vim

So does the GUI. And yes, most users want a different shape.

**Comment by blueyed** [comment]
Related?  https://github.com/neovim/neovim/pull/3165

**Comment by justinmk** [comment]
That's what I was thinking of, thanks.

**Comment by jdebp** [comment]
When the change for #6795 becomes a pull request, please try invoking tmux within Konsole as
```( unset KONSOLE_PROFILE_NAME ; unset KONSOLE_DBUS_SESSION ; tmux )```
and report whether this changes anything.  That _should_ turn off the tmux-wrapped Konsole idiosyncrasy whilst leaving DECSCUSR on.

**Comment by ZyX-I** [comment]
@jdebp Tmux uses environment from the first server startup (plus modifications done via `[tmux ]set-env[ -g]` if any, but this is not relevant), so running tmux like this is not very useful, need to ask tmux users to also provide different socket to start separate server (e.g. `tmux -S "$(mktemp)"`). Not all tmux users are aware of this.

**Comment by jdebp** [comment]
Well whatever magic incantation you use is up to you, although I think that you are overcomplicating things.  Just _start tmux without it or the programs that it in turn runs inheriting those environment variables_, and then report whether this stops it from setting the cursor shape globally.  I am presuming that you already know that you also have to tell tmux how to talk to Konsole about setting the cursor shape with `Ss` and `Se` termcap entries.

**Comment by ZyX-I** [comment]
@jdebp “Just start tmux” does not work if you have it *already started*. Tmux server inherits environment variables *once*, for each next invocation of tmux unless you have previously killed the server environment variables *won’t be inherited*. There is a difference for some environment variables though: e.g. `PATH` is inherited. But `KONSOLE_…` variables are not on the list.

You can always try this yourself: `env -i /usr/bin/tmux -f /dev/null -S /tmp/tmux.sock new-session -d '/usr/bin/env > /tmp/env.cleared ; sleep 5' ; /usr/bin/tmux -f /dev/null -S /tmp/tmux.sock new-session -d '/usr/bin/env > /tmp/env.normal'`. What will appear in the `diff -u /tmp/env.cleared /tmp/env.normal` is just this:

```Patch
--- /tmp/env.cleared	2017-05-25 03:21:05.208520027 +0300
+++ /tmp/env.normal	2017-05-25 03:21:05.208520027 +0300
@@ -1,10 +1,14 @@
+DISPLAY=:0
+PATH=/home/zyx/perl5/bin:/home/zyx/perl5/bin:/home/zyx/.local/bin:/home/zyx/.luarocks/bin:/home/zyx/.gem/ruby/2.2.0/bin:/home/zyx/.gem/ruby/2.1.0/bin:/home/zyx/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/5.4.0:/usr/x86_64-pc-linux-gnu/avr/gcc-bin/5.4.0:/usr/x86_64-pc-linux-gnu/i686-pc-mingw32/gcc-bin/5.4.0:/opt/android-sdk-update-manager/tools:/opt/android-sdk-update-manager/platform-tools:/usr/games/bin:/opt/ekopath/bin
 PWD=/home/zyx/a.a/{tmp}
 TERM=screen
-TMUX=/tmp/tmux.sock,3714,0
-TMUX_PANE=%0
+TMUX=/tmp/tmux.sock,3714,1
+TMUX_PANE=%1
+WINDOWID=20971525
+XAUTHORITY=/home/zyx/.Xauthority
 SHELL=/bin/zsh
 HOME=/home/zyx
 LOGNAME=zyx
-SHLVL=1
+SHLVL=0
 OLDPWD=/home/zyx/a.a/{tmp}
 _=/usr/bin/env
```

But I have run that from Konsole with various KONSOLE_… variables set, have things like LUA_INIT, LESS, etc. All this do *not* appear in `/tmp/env.normal`. If tmux *did* inherit environment variables they would appear in the diff.

**Comment by jdebp** [comment]
So do not have it already started when performing the test.  This is all very simple.  As I said, you are overcomplicating things.  

Build the changed version with the pull request when it comes.  Do _whatever is necessary_ to set up the environment inherited by nvim in the way that I suggested.  Then see whether the behaviour alters.  It's as simple as that.  Subtracting the environment variables should alter the behaviour in a way that makes the cursor change local; and that is the point of the test.  So _figure out  how to do that subtraction_, which presumably you can, and _do it_.

If it does not achieve the desired effect, then try not only subtracting those environment variables but also adding `VTE_VERSION=4000` at the same time.

**Comment by ZyX-I** [comment]
@jdebp I am using tmux for running applications in background. If I kill server I will have a bunch of apps I am using shut down and I am definitely not going to do so. I am not overcomplicating things, I can follow my advice when testing myself without a notice, just suggesting that you should better formulate testing requests for tmux users, I have known some which do not know the detail and https://github.com/neovim/neovim/issues/6778#issuecomment-303862157 was not referencing anybody specific.

BTW, which PR was you referring to? And what specifically is needed to “tell tmux how to talk to Konsole about setting the cursor shape with Ss and Se terminfo entries”, https://github.com/neovim/neovim/issues/4219#issuecomment-182644221?

**Comment by ZyX-I** [comment]
Probably not https://github.com/neovim/neovim/issues/4219#issuecomment-182644221 because either does nothing (without VTE_VERSION=4000, guess it is Neovim which simply does nothing and not tmux) or displays those sequences on screen (worse if `\e` are literal escapes or quotes are changed to double: old bug with popping up sequences).

Full command used:

```Shell
(unset -m KONSOLE_\* ; VTE_VERSION=4000 tmux -S /tmp/tmux.sock -f tmux.conf new \; split 'build/bin/nvim -u NONE -i NONE' \; send-keys -l i)
```

**Comment by justinmk** [comment]
See also #6798 : 

> Instead, a program such as nvim that wants to change the cursor shape has to perform a fake profile update, using a control sequence that first makes a nonce temporary profile with some of the settings changed from the real profile, and that then causes that temporary profile to be applied.
> 
> It is the fact that this done by inventing a Konsole profile and applying it, on the fly, which is causing the effect that you are seeing; where the font size is set to the size specified in the profile. Applying a nonce profile based upon the current one will reset the font size back to what is in the current profile, by the very nature of the action.

**Comment by justinmk** [comment]
Resolution described in https://github.com/neovim/neovim/issues/6798#issuecomment-447152989 . Needs someone to implement it, but we can merge this issue with that one in the meantime.

**Comment by saurabh-kushwah** [comment]
Solved by [#2059](https://github.com/tmux/tmux/issues/2059#issuecomment-575190478)
Worked for me
